src/phase0/state_transition/slot_processing.rs

```diff
@@ -67,11 +75,137 @@ pub fn process_slot<
+    let merkleization_context = context.for_merkleization();
+    let validate_result = validate_result.unwrap_or(true);
+    let block = &signed_block.message;
+
+    process_slots(state, block.slot, context)?;
+    if validate_result {
+        verify_block_signature(state, signed_block, context)?;
+    }
+    process_block(state, block, context)?;
+    if validate_result && block.state_root != state.hash_tree_root(merkleization_context)? {
+        return Err(InvalidStateRoot);
+    }
+
+    Ok(())
+}
+
+pub fn get_matching_source_attestations<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    epoch: Epoch,
+    context: &Context,
+) -> Result<Vec<PendingAttestation<MAX_VALIDATORS_PER_COMMITTEE>>, Error> {
+    let previous_epoch = get_previous_epoch(state, context);
+    let current_epoch = get_current_epoch(state, context);
+
+    if previous_epoch != epoch && current_epoch != epoch {
+        return Err(InvalidEpoch {
+            previous: previous_epoch,
+            current: current_epoch,
+        });
+    }
+
+    if epoch == current_epoch {
+        return Ok(state.current_epoch_attestations.to_vec());
+    }
+    Ok(state.previous_epoch_attestations.to_vec())
+}
+
+pub fn get_matching_target_attestations<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    epoch: Epoch,
+    context: &Context,
+) -> Result<Vec<PendingAttestation<MAX_VALIDATORS_PER_COMMITTEE>>, Error> {
+    let source_attestations = get_matching_source_attestations(state, epoch, context)?;
+    let block_root = get_block_root(state, epoch, context)?;
+    Ok(source_attestations
+        .into_iter()
+        .filter(|a| a.data.target.root.as_ref() == block_root.as_ref())
+        .collect())
+}
```

Comment:
i'd suggest dropping the intermediate allocations here, unless the ergonomics just gets radically better...

in this case, we can just return the existential type `impl Iterator`

```rust
get_matching_target_attestations<...>(state, epoch, context) -> Result<impl Iterator<Item=PendingAttestation<...>>, Errror>
```

and it may be worth adjusting the type signatures so they are more readable but that is something we can do later....


if we make this change then the last line would be:

```rust
Ok(source_attestations.into_iter().filter(|a| ...))
```